app/models/sample.rb

```diff
@@ -387,6 +387,16 @@ def self.editable(user)
     end
   end
 
+  def deletable?(user)
+    if user.admin?
+      true
+    elsif user_id == user.id
+      # Sample belongs to the user
+      # Allow deletion if no pipeline runs, or report failed.
+      pipeline_runs.empty? || pipeline_runs[0].report_failed?
```

Comment:
loop through the pipeline_runs and get the pipeline_runs that didn't fail and make sure it's not empty?